corehq/apps/ota/views.py

```diff
@@ -168,26 +169,29 @@ def get_restore_response(domain, couch_user, app_id=None, since=None, version='1
         )
         return HttpResponse(response, content_type="text/xml; charset=utf-8", status=412), None
 
-    # not a view just a view util
+    is_demo_restore = couch_user.is_commcare_user() and couch_user.is_demo_user
+    if is_demo_restore:
+        # if user is in demo-mode, return demo restore
+        update_device_id(couch_user, device_id)
+        return demo_user_restore_response(couch_user), None
+
+    uses_login_as = bool(as_user)
+    as_user_obj = CouchUser.get_by_username(as_user) if uses_login_as else None
+    if uses_login_as and not as_user_obj:
+        msg = _(u'Invalid restore as user {}').format(as_user)
+        return HttpResponse(msg, status=401), None
     is_permitted, message = is_permitted_to_restore(
         domain,
         couch_user,
-        as_user,
+        as_user_obj,
     )
     if not is_permitted:
         return HttpResponse(message, status=401), None
 
-    is_demo_restore = couch_user.is_commcare_user() and couch_user.is_demo_user
-    couch_restore_user = couch_user
-    if not is_demo_restore and as_user is not None:
-        couch_restore_user = CouchUser.get_by_username(as_user)
+    couch_restore_user = as_user_obj if uses_login_as else couch_user
     update_device_id(couch_restore_user, device_id)
```

Comment:
This function updates properties in `user_data` that need to be current for the restore.  I think that the calls to `CouchUser.get_by_username(as_user)` might've been hitting a cache or something and getting out-of-date information.  This PR passes around the same user object to try to avoid that.

Reply:
is this the reason for this PR? the `get_by_username` cache is cleared on user save

Reply:
It is.  I could see this being a race condition with multiple sessions hitting this at the same time.  Passing around the same user object should make sure that the version used in restore generation is the one updated here.

Reply:
(we discussed offline but a lighter-weight way to fix race conditions is to just use `CriticalSection`, and identify the smallest block of code necessary to be the critical section)

Reply:
@ctsims I think this should solve the issue, since the device id info is being set on the this user object, which is passed around and eventually used in the restore generation (ie, it's not later refreshed from the DB).  That still feels a little sketchy though.  @gcapalbo is proposing locking the restore so there can be only one restore at a time for a given username.  Do you think that would be an acceptable approach?  If restores are still slow, that could mean waiting to aquire a lock.  Any sense of how big an issue that might be?

Reply:
I think locking per user should be fine for eNikshay for now if it's hard to fix more generally. Currently our restores are quite fast, and the web apps users almost never login-as the same user at once for logistical reasons.

Locking per user would be way rougher on some other projects which use a very limited number of user accounts, though

Reply:
My preference would be to merge this PR and then separately look into locking this part of the restore (since I'd want this change regardless).  @gcapalbo how does that sound to you?

Reply:
i'm fine with the change though i don't have much general context about the restore process. it might be worth getting sign off from someone more familiar with restores as well.

if we do release it, it should be tested - has it been?